scripts/create_changelog.py

```diff
@@ -0,0 +1,343 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+This script is used in 'Generate Changelog' workflow, defined in the 'generate_changelog.yml' file
+to automate the process of creating changelogs after each stable release.
+
+It fetches release data, extracts relevant pull request (PR) information, and generates a detailed changelog
+and add it to src/changelogs.
+The script uses the GitHub REST API to collect information about contributors, PR authors, and reviewers.
+
+Additional Note:
+- The script requires a GitHub personal access token (PAT) stored in an environment variable named `GITHUB_PAT`.
+"""
+
+import requests
+import re
+from dotenv import load_dotenv
+import os
+import argparse
+import xml.etree.ElementTree as ET
+from unidecode import unidecode
+import html
+
+load_dotenv()
+GITHUB_PAT = os.getenv('GITHUB_PAT')
+BASE_URL = r"https://api.github.com/repos/sagemath/sage"
+HEADERS = {'Authorization': f'token {GITHUB_PAT}', }
+AUTOMATED_BOTS = ['dependabot[bot]', 'github-actions', 'renovate[bot]']
+
+# Maps the github username to contributer name
+git_to_name = {}
+
+# Stores unique names of contributors
+all_contribs = set([])
+
+# Stores unique names of contributors
+first_contribs = set([])
+
+# Stores information for all the prs across all pre-releases
+# Maps tag of pre-release to its info
+all_info = {}
+
+# Stores new names which didn't previously existed in conf/contributor.xml
+new_names = []
+
+
+def get_last_name(full_name):
+    return full_name.split()[-1]
+
+
+def merge_contributors(original_file, new_contributors):
+    with open(original_file, 'r',encoding='utf-8') as f:
+        original_content = f.read()
+    tree = ET.fromstring(original_content)
+
+    for new_name, github in new_contributors:
+        new_contributor = ET.Element('contributor')
+        new_contributor.set('name', new_name)
+        new_contributor.set('github', github)
+
+        tree.append(new_contributor)
+    
+    sorted_contributors = sorted(
+        tree.findall('contributor'),
+        # Items which don't have a name attribute are placed at the end 
+        key=lambda x: unidecode(get_last_name(x.get('name', '\x7F'))) # '\x7F' is highest ASCII value
+    )
+
+    for cont in tree.findall('contributor'):
+        tree.remove(cont)
+
+    for cont in sorted_contributors:
+        tree.append(cont)
+
```

Comment:
This sorts the list of contributors by "last name" all again each time when a release is made. This may destroy the correct order of some contributors made manually since their last names are special.

How about inserting each of new contributors into the tree finding the correct place while keeping the order of existing contributors? This would be a bit more efficient as well (though efficiency is a least concern here...)

You may use the same sort key `key=lambda x: unidecode(get_last_name(x.get('name', '\x7F'))` to find the right place for each contributor.

Reply:
I did consider this approach. But this approach seemed to have a significant limitation.
For example, consider the following existing order in `contributors.xml`
```
<contributor name="John Doe"/>
<contributor name="Jane Ann Doe Zander"/>
..
..
```
Here, the actual last name of the second contributor is not the final word `Zander` but the third word `Doe`
If following contributors were to be added to the list:
- "Alice Henderson"
- "Bob Johnson"
- "Michael Peterson"
- "Paul Robertson"

The insertion algorithm would place all the new contributors between the two existing contributors, as shown below:
```
<contributor name="John Doe"/>
<contributor name="Alice Henderson"/>
<contributor name="Bob Johnson"/>
<contributor name="Michael Peterson"/>
<contributor name="Paul Robertson"/>
<contributor name="Jane Ann Doe Zander"/>
..
..
```
This placement is incorrect for almost all the new names, as they end up in the wrong part of the list. Furthermore, this could create a cascade effect (if the order is not mannually corrected).

Any new contributors with similarly complex last-name structures (as Jane Ann Doe Zander) added in the future might make the insertion unreliable. 
 It might repeatedly misplace entries, ultimately defeating the purpose of maintaining a correct order.

Reply:
Right. That is also awful.

Let's go with your current approach.